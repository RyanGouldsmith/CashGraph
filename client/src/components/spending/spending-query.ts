import gql from 'graphql-tag';

export const GetSpendingQuery = gql`
  query GetSpending($userId: String!, $limit: Int) {
    spending(userId: $userId, limit: $limit) {
      title
      price
    }
  }
`;
