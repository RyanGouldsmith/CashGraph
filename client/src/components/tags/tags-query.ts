import gql from 'graphql-tag';

export const GetTagsQuery = gql`
  query GetTags($userId: String!) {
    tags(userId: $userId) {
      id
      name
      colour
    }
  }
`;

gql`
  type TagInput {
    name: TagName
    colour: TagColour
  }
`;

gql`
  enum TagName {
    HOLIDAY
    FOOD
    TRAVEL
    ENTERTAINMENT
  }
`;

gql`
  enum TagColour {
    GREEN
    RED
    BLUE
  }
`;
