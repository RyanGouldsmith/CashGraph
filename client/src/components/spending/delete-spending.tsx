import React from 'react';
import { UserProvider } from '../user/user-provider';
import { Mutation, MutationFn, MutationResult } from 'react-apollo';
import { withRouter, RouteComponentProps } from 'react-router';
import { Link } from 'react-router-dom';

import { DeleteSpendingMutation } from './spending-query';
import { DeleteSpendingResult } from './spending-types';

const SpendingDelete: React.FunctionComponent<RouteComponentProps<any>> = ({
  match,
  history
}) => {
  const { id } = match.params;

  const goBack = () => history.push('/');
  const handleDelete = (deleteSpending: MutationFn, userId: string) => (
    event: React.MouseEvent<HTMLButtonElement>
  ) => {
    event.stopPropagation();
    const variables = {
      spending: {
        userId,
        id
      }
    };
    deleteSpending({ variables });
  };

  return (
    <UserProvider.Consumer>
      {(userId: string) => (
        <React.Fragment>
          <h2>Are you sure you want to delete this item?</h2>
          <Mutation mutation={DeleteSpendingMutation}>
            {(
              deleteSpending: MutationFn,
              result: MutationResult<DeleteSpendingResult>
            ) => (
              <React.Fragment>
                {result.data ? (
                  <React.Fragment>
                    <h1>Deleted Successully</h1>
                    <Link to="/">Go Back</Link>
                  </React.Fragment>
                ) : (
                  <section className="spending__delete">
                    <button onClick={goBack}>No, go to Dashboard</button>
                    <button onClick={handleDelete(deleteSpending, userId)}>
                      Yes Delete
                    </button>
                  </section>
                )}
              </React.Fragment>
            )}
          </Mutation>
        </React.Fragment>
      )}
    </UserProvider.Consumer>
  );
};

const SpendingDeleteWithRouter = withRouter(SpendingDelete);

export default SpendingDeleteWithRouter;
