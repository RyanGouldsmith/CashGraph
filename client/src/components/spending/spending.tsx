import React from 'react';
import { Query } from 'react-apollo';
import { SpendingType, SpendingQueryResult } from './spending-types';
import { GetSpendingQuery } from './spending-query';
import { UserProvider } from '../user/user-provider';
import { Loading } from '../loading/loading';
import { Link } from 'react-router-dom';

const totalSpending = (spendings: Array<SpendingType>): number => {
  return spendings.reduce((acc, curr) => {
    return (acc += curr.price);
  }, 0.0);
};

interface SpendingResult {
  loading: boolean;
  data: SpendingQueryResult;
}

interface SpendingProps {
  limit?: Number;
  shouldShowEditLink?: Boolean;
  shouldShowDeleteLink?: Boolean;
}

export const Spending: React.SFC<SpendingProps> = ({
  limit,
  shouldShowEditLink = true,
  shouldShowDeleteLink = true,
}) => {
  return (
    <UserProvider.Consumer>
      {userId => (
        <React.Fragment>
          <h1>Spending ... </h1>
          <Query query={GetSpendingQuery} variables={{ userId, limit }}>
            {({ loading, data }: SpendingResult) => {
              if (loading) return <Loading />;
              if (data) {
                const { spending } = data;
                return (
                  <section className="spending">
                    <p>{`Total Spending is Â£${totalSpending(spending).toFixed(2)}`}</p>
                    {spending.map((item: SpendingType) => {
                      return (
                        <section className="spending__item">
                          <p className="spending__item-title" key={item.title}>
                            {item.title}
                          </p>
                          {shouldShowEditLink && (
                            <Link to={`/spending/edit/${item.id}`}>Edit Spending Item</Link>
                          )}
                          {shouldShowDeleteLink && (
                            <Link to={`/spending/delete/${item.id}`}>Delete Spending Item</Link>
                          )}
                        </section>
                      );
                    })}
                  </section>
                );
              }
              return null;
            }}
          </Query>
        </React.Fragment>
      )}
    </UserProvider.Consumer>
  );
};

export default Spending;
