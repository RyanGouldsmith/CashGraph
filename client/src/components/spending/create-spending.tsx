import React from 'react';
import { Mutation, MutationFn, MutationResult, Query } from 'react-apollo';

import { UserProvider } from '../user/user-provider';
import { GetTagsQuery } from '../tags/tags-query';
import { Loading } from '../loading/loading';

import { SpendingMutation } from './spending-query';
import { SpendingForm } from './spending-form';

export const CreateSpending: React.FunctionComponent<{}> = () => {
  let titleRef = React.createRef<HTMLInputElement>();
  let priceRef = React.createRef<HTMLInputElement>();
  let tagRef = React.createRef<HTMLSelectElement>();

  const submitForm = (mutationHandler: MutationFn, userId: String) => (
    event: React.FormEvent<HTMLFormElement>
  ) => {
    event.preventDefault();
    const variables = {
      spending: {
        title: titleRef.current.value,
        price: Number.parseFloat(priceRef.current.value),
        tag: {
          id: tagRef.current.value
        },
        userId
      }
    };
    mutationHandler({ variables });
  };

  return (
    <UserProvider.Consumer>
      {userId => (
        <Query query={GetTagsQuery} variables={{ userId }}>
          {({ loading, data }) => {
            if (loading) return <Loading />;
            if (data) {
              const { tags } = data;
              return (
                <Mutation mutation={SpendingMutation}>
                  {(createSpending: MutationFn, result: MutationResult) => (
                    <React.Fragment>
                      {result.data ? (
                        <h1>Created successfully</h1>
                      ) : (
                        <SpendingForm
                          submitForm={submitForm}
                          spendingCallback={createSpending}
                          userId={userId}
                          titleRef={titleRef}
                          priceRef={priceRef}
                          tagRef={tagRef}
                          spendingData={{ tags }}
                        />
                      )}
                    </React.Fragment>
                  )}
                </Mutation>
              );
            }
            return null;
          }}
        </Query>
      )}
    </UserProvider.Consumer>
  );
};

export default CreateSpending;
