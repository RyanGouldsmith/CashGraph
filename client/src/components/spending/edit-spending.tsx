import React from 'react';
import { UserProvider } from '../user/user-provider';
import { Mutation, MutationFn, MutationResult, Query } from 'react-apollo';
import { withRouter, RouteComponentProps } from 'react-router';

import { Loading } from '../loading/loading';
import { GetTagsQuery } from '../tags/tags-query';

import { EditSpendingMutation, GetSpendingItem } from './spending-query';
import { SpendingItemType } from './spending-types';
import { SpendingForm } from './spending-form';

type SpendingItemResult = {
  loading: boolean;
  data: SpendingItemType;
};

const EditSpending: React.FunctionComponent<RouteComponentProps<any>> = ({
  match
}) => {
  let titleRef = React.createRef<HTMLInputElement>();
  let priceRef = React.createRef<HTMLInputElement>();
  let tagRef = React.createRef<HTMLSelectElement>();
  const id = match.params.id;

  const submitForm = (mutationHandler: MutationFn, userId: String) => (
    event: React.FormEvent<HTMLFormElement>
  ) => {
    event.preventDefault();
    const variables = {
      spending: {
        title: titleRef.current.value,
        price: Number.parseFloat(priceRef.current.value),
        tag: {
          id: tagRef.current.value
        },
        userId,
        id
      }
    };
    mutationHandler({ variables });
  };

  return (
    <UserProvider.Consumer>
      {(userId: string) => (
        <React.Fragment>
          <h1>Edit Spending</h1>
          <Query query={GetTagsQuery} variables={{ userId }}>
            {({ loading, data }) => {
              if (loading) return <Loading />;
              if (data) {
                const { tags } = data;
                return (
                  <Query
                    query={GetSpendingItem}
                    variables={{ spending: { userId, id } }}
                  >
                    {({ loading, data }: SpendingItemResult) => {
                      if (loading) return <Loading />;
                      if (data) {
                        const { title, price, tag } = data.getSpendingItem;
                        return (
                          <Mutation mutation={EditSpendingMutation}>
                            {(
                              editSpending: MutationFn,
                              result: MutationResult
                            ) => (
                              <React.Fragment>
                                {result.data ? (
                                  <h1>Edited successfully</h1>
                                ) : (
                                  <SpendingForm
                                    submitForm={submitForm}
                                    spendingCallback={editSpending}
                                    userId={userId}
                                    titleRef={titleRef}
                                    priceRef={priceRef}
                                    tagRef={tagRef}
                                    spendingData={{
                                      title,
                                      price,
                                      tags,
                                      selectedTag: tag[0].id
                                    }}
                                  />
                                )}
                              </React.Fragment>
                            )}
                          </Mutation>
                        );
                      }
                      return null;
                    }}
                  </Query>
                );
              }
              return null;
            }}
          </Query>
        </React.Fragment>
      )}
    </UserProvider.Consumer>
  );
};

const EditSpendingWithRouter = withRouter(EditSpending);

export default EditSpendingWithRouter;
