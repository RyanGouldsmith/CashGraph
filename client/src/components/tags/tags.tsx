import React from 'react';
import { Query } from 'react-apollo';

import { GetTagsQuery } from './tags-query';
import { TagType } from './tags-types';
import { Loading } from '../loading/loading';

export const Tags: React.SFC<{}> = _ => {
  return (
    <React.Fragment>
      <h1>The tags are: </h1>
      <Query query={GetTagsQuery}>
        {({ loading, data }) => {
          if (loading) return <Loading />;
          if (data) {
            const { tags } = data;
            return (
              <div className="tags">
                {tags.map((tag: TagType) => {
                  return (
                    <p style={{ color: `${tag.colour}` }} className="tags__tag" key={tag.name}>
                      {tag.name}
                    </p>
                  );
                })}
              </div>
            );
          }
          return null;
        }}
      </Query>
    </React.Fragment>
  );
};
export default Tags;
