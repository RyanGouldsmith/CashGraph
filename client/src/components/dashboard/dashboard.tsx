import React from 'react';

import { Tags } from '../tags/tags';
import { Spending } from '../spending/spending';
import { Link } from 'react-router-dom';

export const Dashboard: React.SFC<{}> = () => {
  return (
    <React.Fragment>
      <h1>Cash Graph Dashboard</h1>
      <main className="dashboard">
        <section className="dashboard__tags">
          <Tags />
        </section>
        <section className="dashboard__spending">
          <Link to="/spending">More Spending ...</Link>
          <Spending limit={2} />
        </section>
      </main>
    </React.Fragment>
  );
};
