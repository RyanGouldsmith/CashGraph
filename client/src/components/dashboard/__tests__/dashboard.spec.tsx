import 'babel-polyfill';
import React from 'react';
import { render, cleanup, wait } from 'react-testing-library';
import { MockedProvider } from 'react-apollo/test-utils';
import { BrowserRouter } from 'react-router-dom';
import { GetSpendingQuery } from '../../spending/spending-query';
import { GetTagsQuery } from '../../tags/tags-query';
import { Dashboard } from '../dashboard';

afterEach(cleanup);

const mocks = [
  {
    request: {
      query: GetSpendingQuery,
      variables: { limit: 2 },
    },
    result: {
      data: {
        spending: [
          { title: 'testSpendingTitle', price: 3.0 },
          { title: 'anotherSpendingTitle', price: 7.0 },
        ],
      },
    },
  },
  {
    request: {
      query: GetTagsQuery,
    },
    result: {
      data: {
        tags: [
          {
            name: 'ENTERTAINMENT',
            colour: 'RED',
          },
          {
            name: 'TRAVEL',
            colour: 'BLUE',
          },
        ],
      },
    },
  },
];

test('should match the snapshot containing tags and spending', async () => {
  const { container } = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <BrowserRouter>
        <Dashboard />
      </BrowserRouter>
    </MockedProvider>,
  );

  await wait(() => {
    expect(container).toMatchSnapshot();
  });
});
