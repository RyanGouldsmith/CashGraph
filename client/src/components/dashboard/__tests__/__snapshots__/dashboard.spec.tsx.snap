// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot containing tags and spending 1`] = `
<div>
  <h1>
    Cash Graph Dashboard
  </h1>
  <main
    class="dashboard"
  >
    <section
      class="dashboard__tags"
    >
      <h1>
        The tags are: 
      </h1>
      <div
        class="tags"
      >
        <p
          class="tags__tag"
          style="color: RED;"
        >
          ENTERTAINMENT
        </p>
        <p
          class="tags__tag"
          style="color: BLUE;"
        >
          TRAVEL
        </p>
      </div>
    </section>
    <section
      class="dashboard__spending"
    >
      <a
        href="/spending"
      >
        More Spending ...
      </a>
      <h1>
        Spending ... 
      </h1>
      <section
        class="spending"
      >
        <p>
          Total Spending is Â£10
        </p>
        <p
          class="spending__title"
        >
          testSpendingTitle
        </p>
        <p
          class="spending__title"
        >
          anotherSpendingTitle
        </p>
      </section>
    </section>
  </main>
</div>
`;
